### PR TITLE
Add calculation for tridiagonal matrices (solve, factorize, det, rcond)

### PR DESCRIPTION
**Pull request by doraneko94**
I added the calculation for tridiagonal matrices, which might be important for solving differential equations.

First, I made a struct of `TriDiagonal` for LU factorization and solving equation using LAPACK. It represents a tridiagonal matrix as 3 one-dimensional `Array` (diagonal, sub-diagonal and super-diagonal elements). 
And this struct also holds the layout and 1-norm of the raw matrix. They are used in some methods (`solve_tridiagonal`, `rcond_tridiagonal`, etc.)

Second, `LUFactorizedTriDiagonal` struct contains additional 1D `Array` of the second super-diagonal of matrix U and `ipiv`, just like `LUFactorized`.

These structs and `ArrayBase<S, Ix2>` have `lu (*gttrf)`, `rcond (*gtcon)`, `solve (*gttrs)` like methods using LAPACK and a `det` like method using a recurrent relation implemented in Rust.

I also added an example and some test functions. Please confirm that you can use them just like any other methods implemented for  general or triangular matrices.

### CODE REVIEW

**Comment by doraneko94** [comment]
I removed a 1-norm field `n1`  from `Tridiagonal` and add it to `LUFactorizedTridiagonal` (with renaming to `anom` ).
This is because I wanted to implement `Index` / `IndexMut` trait to `Tridiagonal` so that it can be updated the value by indexing.
Thus, when you call `.lu_tridiagonal()`, 1-norm is calculated too to construct a `LUFactorizedTridiagonal`.

File: src/tridiagonal.rs [hunk 1]
```diff
@@ -0,0 +1,801 @@
...
+#[derive(Clone)]
+pub struct LUFactorizedTriDiagonal<A: Scalar> {
+    /// A tridiagonal matrix which consists of
+    /// - l : layout of raw matrix
+    /// - dl: (n-1) multipliers that define the matrix L.
+    /// - d : (n) diagonal elements of the upper triangular matrix U.
+    /// - du: (n-1) elements of the first super-diagonal of U.
+    pub a: TriDiagonal<A>,
+    /// (n-2) elements of the second super-diagonal of U.
+    pub du2: Array1<A>,
```
Comment (termoshtt) [inline src/tridiagonal.rs:197]: Please use `Vec<A>` as above

File: src/tridiagonal.rs [hunk 2]
```diff
@@ -0,0 +1,801 @@
...
+impl<A, S> ToTriDiagonal<A> for ArrayBase<S, Ix2>
+where
+    A: Scalar + Lapack,
+    S: Data<Elem = A>,
+{
+    fn to_tridiagonal(&self) -> Result<TriDiagonal<A>> {
+        let l = self.square_layout()?;
+        let (n, _) = l.size();
+        if n < 2 {
+            panic!("Cannot make a tridiagonal matrix of shape=(1, 1)!");
```
Comment (termoshtt) [inline src/tridiagonal.rs:114]: Should not panic here. Please define an Error variant and throw it.

File: src/tridiagonal.rs [hunk 3]
```diff
@@ -0,0 +1,801 @@
...
+        }
+    }
+}
+
+/// An interface for making a TriDiagonal struct.
+pub trait ToTriDiagonal<A: Scalar> {
+    /// Extract tridiagonal elements and layout of the raw matrix.
+    ///
+    /// If the raw matrix has some non-tridiagonal elements,
+    /// they will be ignored.
```
Comment (termoshtt) [inline src/tridiagonal.rs:99]: I think `to_tridiagonal_unchecked` might be better in order to emphasize the non-tridiagonal elements are dropped. Or use more explicit verb, e.g. `extract_tridiagonal`

File: src/tridiagonal.rs [hunk 4]
```diff
@@ -0,0 +1,801 @@
...
+
+use super::convert::*;
+use super::error::*;
+use super::lapack::*;
+use super::layout::*;
+
+/// Represents a tridiagonal matrix as 3 one-dimensional vectors.
+/// This struct also holds the layout of the raw matrix.
+#[derive(Clone, PartialEq)]
+pub struct TriDiagonal<A: Scalar> {
```
Comment (termoshtt) [inline src/tridiagonal.rs:19]:
```suggestion
pub struct Tridiagonal<A: Scalar> {
```

"Tridiagonal" seems not to be a compound word

- https://mathworld.wolfram.com/TridiagonalMatrix.html
- https://en.wikipedia.org/wiki/Tridiagonal_matrix

File: src/tridiagonal.rs [hunk 5]
```diff
@@ -0,0 +1,801 @@
...
+    /// layout of raw matrix
+    pub l: MatrixLayout,
+    /// (n-1) sub-diagonal elements of matrix.
+    pub dl: Array1<A>,
+    /// (n) diagonal elements of matrix.
+    pub d: Array1<A>,
+    /// (n-1) super-diagonal elements of matrix.
+    pub du: Array1<A>,
```
[the remark this answers is not in the document]
Comment (doraneko94) [reply src/tridiagonal.rs]: Fixed.

File: src/tridiagonal.rs [hunk 6]
```diff
@@ -0,0 +1,801 @@
...
+
+fn debug_bounds_check_tridiag(n: i32, row: i32, col: i32) {
+    if std::cmp::max(row, col) >= n {
+        panic!(
+            "ndarray: index {:?} is out of bounds for array of shape {}",
+            [row, col],
+            n
+        );
+    }
+}
```
[the remark this answers is not in the document]
Comment (doraneko94) [reply src/tridiagonal.rs]: Fixed.

File: src/tridiagonal.rs (hunk 1 above)
Comment (doraneko94) [reply src/tridiagonal.rs]: Fixed.

File: src/tridiagonal.rs (hunk 2 above)
Comment (doraneko94) [reply src/tridiagonal.rs]: I defined a new `LinalgError` named `NotStandardShape` in consideration of versatility.

File: src/tridiagonal.rs (hunk 3 above)
Comment (doraneko94) [reply src/tridiagonal.rs:99]: I prefer `extract_tridiagonal`.

File: src/tridiagonal.rs (hunk 4 above)
Comment (doraneko94) [reply src/tridiagonal.rs]: I renamed all the part of `TriDiagonal` to `Tridiagonal`.

File: src/tridiagonal.rs [hunk 7]
```diff
@@ -0,0 +1,805 @@
...
+
+pub trait TridiagIndex {
+    fn to_tuple(&self) -> (i32, i32);
+}
+impl TridiagIndex for [Ix; 2] {
+    fn to_tuple(&self) -> (i32, i32) {
+        (self[0] as i32, self[1] as i32)
+    }
+}
```
Comment (termoshtt) [inline src/tridiagonal.rs:37]: I think [`ndarray::NdIndex<Ix2>`](https://docs.rs/ndarray/0.13.1/ndarray/trait.NdIndex.html) could be better for it, but it does not enough iterface for implementing this PR, e.g. `Into<Ix2>`. I keep this code as it is.

**Comment by termoshtt** [comment]
Thank a lot :)

**Pull request merged**